lib/runCall.js

```diff
@@ -59,111 +61,279 @@ module.exports = function (opts, cb) {
-  client.on('data', function (data) {
-    try {
-      var query = msg_pb.VMQuery.deserializeBinary(data)
+  function checkpointState (_callback) {
+    stateManager.checkpoint(_callback)
+  }
+
+  function runClient (_callback) {
+    // Add a 'data' event handler for the client socket
+    // data is what the server sent to this socket
+    client.on('data', function (data) {
+      var query
+      try {
+        query = msg_pb.VMQuery.deserializeBinary(data)
+      } catch (err) {
+        if (err.message === 'Assertion failed') {
+          var shortData = new Uint8Array(Buffer.from(data).slice(4))
+          query = msg_pb.VMQuery.deserializeBinary(shortData)
+        }
+      }
       switch (query.getQueryCase()) {
         case GETACCOUNT: {
-          console.log('GETACCOUNT', data)
           computeGetAccount(query)
           break
         }
         case GETSTORAGEDATA: {
-          console.log('GETSTORAGEDATA', data)
           computeGetStorageData(query)
           break
         }
         case GETCODE: {
-          console.log('GETCODE', data)
-          // TODO
+          computeGetCode(query)
           break
         }
         case GETBLOCKHASH: {
-          console.log('GETBLOCKHASH', data)
           // TODO
           break
         }
+        case CALLRESULT: {
+          console.log('CALLRESULT')
+          client.destroy()
+          computeCallResult(query, _callback)
+          break
+        }
+        case QUERY_NOT_SET: {
+          console.log('QUERYNOTSET')
+          break
+        }
       }
-    } catch (err) {
-      var mdata = new Uint8Array(data.length - 4)
-      for (var i = 4; i < data.length; i++) {
-        mdata[i - 4] = data[i]
+    })
+
+    // Add a 'close' event handler for the client socket
+    client.on('close', function () {
+      console.log('Connection closed')
+    })
+
+    client.connect(kevmPort, kevmHost, function () {
+      console.log('CONNECTED TO: ' + kevmHost + ':' + kevmPort)
+      var hello = createHello()
+      client.write(hello)
+      var callCtx = createCallContext()
+      client.write(callCtx)
+    })
+  }
+
+  function computeCallResult (query, _callback) {
+    var gasProvided = new BN(block.transactions[0].gas)
+
+    var callResultObject = query.getCallresult()
+    var returnData = callResultObject.getReturndata()
+    returnData = returnData.length ? ethUtil.bufferToHex(Buffer.from(returnData)) : returnData
+    var returnCode = callResultObject.getReturncode()
+    var gasRemaining = new BN(callResultObject.getGasremaining())
+    var gasRefund = new BN(callResultObject.getGasrefund())
+    var error = callResultObject.getError()
+    var modifiedAccounts = callResultObject.getModifiedaccountsList()
+    var logEntries = callResultObject.getLogsList()
+    var logs = fromLogEntries(logEntries)
+    if (returnCode.length === 20) {
+      createdAddress = ethUtil.bufferToHex(Buffer.from(returnCode))
+    } else if (returnCode.length === 21) {
```

Comment:
The `createdAddress` is now taken from the `CallResult` message. There are cases in which the `returnCode` has an extra byte on the first position which results in an invalid address. I'm not sure why it happens and this should be considered as a temporary work around.